app/lib/lets-encrypt.server.ts

```diff
@@ -10,6 +12,62 @@ import type {
 } from 'acme-client/types/rfc8555';
 import { secrets } from 'docker-secret';
 
+/**
+ * Get the zone domain (i.e., drop subdomains that are not in the root level of the zone)
+ * @param {string} domain
+ * @returns {Promise<string>} domain
+ */
+const getHostedZoneForDomain = async (incomingDomain: string): Promise<string> => {
+  let domainName = incomingDomain;
+  let found = false;
+
+  // loop until we find a soa, each time drop a subdomain from `domainName`
+  while (domainName.includes('.') && !found) {
+    try {
+      await dnsPromises.resolveSoa(domainName);
+      found = true;
+    } catch (e) {
+      // remove one level, i.e., c.b.a.com => b.a.com
+      domainName = domainName.replace(/^.*?\./, '');
```

Comment:
Why the `?` after `*` here?

Would it not be better to use `/^[^.]*\./`?

Reply:
The ? makes * to not be "greedy", so it will match as few characters as possible, making the subsequent \. be the first dot in the string (if there was no `?` * would gulp up everything until the last '.' in the domain).

The effect is the same to what you suggested

Reply:
Sounds good, thanks for the explanation.